File: mobile/modules/aegis-sodium/android/src/main/java/expo/modules/aegissodium/VaultKek.kt
package expo.modules.aegissodium

import android.content.Context
import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.security.keystore.StrongBoxUnavailableException
import android.util.Base64
import expo.modules.kotlin.exception.CodedException
import java.security.KeyStore
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * F-1b key vault: the per-profile key-encryption key (KEK) that wraps every
 * private key the C vault hands out as a blob (docs/F1B-KEY-VAULT-DESIGN.md).
 *
 * Each KEK is 32 random bytes, stored in the app's private preferences
 * encrypted (AES-256-GCM) under one NON-EXPORTABLE Android Keystore key
 * (StrongBox when the device has one). JavaScript never sees a KEK: it goes
 * from here straight into the C vault's guarded memory, and this copy is zeroed.
 *
 * If the Keystore key is gone (keystore reset, app data restored onto another
 * device), the stored KEK cannot be decrypted: ERR_AEGIS_VAULT_KEK_LOST, never
 * a silently regenerated KEK (that would orphan every blob without saying so).
 */
internal object VaultKek {
  private const val KEYSTORE = "AndroidKeyStore"
  private const val WRAP_ALIAS = "aegis_vault_kek_wrap_v1"
  private const val PREFS = "aegis_vault_v1"
  private const val GCM_IV = 12
  private const val GCM_TAG_BITS = 128

  private fun prefKey(slot: String) = "kek_$slot"

  private fun wrapKey(create: Boolean): SecretKey? {
    val ks = KeyStore.getInstance(KEYSTORE).apply { load(null) }
    (ks.getKey(WRAP_ALIAS, null) as? SecretKey)?.let { return it }
    if (!create) return null
    val gen = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE)
    fun spec(strongBox: Boolean) =
      KeyGenParameterSpec.Builder(WRAP_ALIAS, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
        .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
        .setKeySize(256)
        .setRandomizedEncryptionRequired(true)
        .apply { if (strongBox && Build.VERSION.SDK_INT >= 28) setIsStrongBoxBacked(true) }
        .build()
    if (Build.VERSION.SDK_INT >= 28) {
      try {
        gen.init(spec(true))
        return gen.generateKey()
      } catch (_: StrongBoxUnavailableException) {
        // No secure element: the TEE-backed Keystore below.
      }
    }
    gen.init(spec(false))
    return gen.generateKey()
  }

  /** The profile's KEK, created on first use. The caller zeroes the returned array. */
  fun kek(context: Context, slot: String): ByteArray {
    val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    val stored = prefs.getString(prefKey(slot), null)
    if (stored != null) {
      val key = wrapKey(create = false) ?: throw lost("the vault's Keystore key is missing")
      val raw = Base64.decode(stored, Base64.NO_WRAP)
      if (raw.size <= GCM_IV) throw lost("stored KEK is malformed")
      return try {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_BITS, raw, 0, GCM_IV))
        cipher.doFinal(raw, GCM_IV, raw.size - GCM_IV)
      } catch (e: Exception) {
        throw lost("stored KEK does not decrypt: ${e.javaClass.simpleName}")
      }
    }
    val kek = ByteArray(32).also { SecureRandom().nextBytes(it) }
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, wrapKey(create = true))
    val ct = cipher.doFinal(kek)
    val out = cipher.iv + ct
    if (!prefs.edit().putString(prefKey(slot), Base64.encodeToString(out, Base64.NO_WRAP)).commit()) {
      kek.fill(0)
      throw CodedException("ERR_AEGIS_VAULT", "could not persist the vault KEK", null)
    }
    return kek
  }

  /** Forget the profile's KEK: its blobs can never be unwrapped again (cryptographic erase). */
  fun destroy(context: Context, slot: String) {
    val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    if (!prefs.edit().remove(prefKey(slot)).commit()) {
      throw CodedException("ERR_AEGIS_VAULT", "could not delete the vault KEK", null)
    }
  }

  private fun lost(why: String) = CodedException("ERR_AEGIS_VAULT_KEK_LOST", why, null)
}

File: mobile/modules/aegis-sodium/ios/VaultKek.swift
import ExpoModulesCore
import Foundation
import Security

/// F-1b key vault: the per-profile key-encryption key (KEK) that wraps every
/// private key the C vault hands out as a blob (docs/F1B-KEY-VAULT-DESIGN.md).
///
/// Each KEK is 32 random bytes in the Keychain, readable after the first
/// unlock and never migrated off this device (not in iCloud or encrypted
/// backups). JavaScript never sees it: it goes from here into the C vault's
/// guarded memory, and this copy is zeroed.
enum VaultKek {
  private static let service = "org.aegislink.vault.kek.v1"

  private static func query(_ slot: String) -> [String: Any] {
    [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: slot,
    ]
  }

  /// The profile's KEK, created on first use. The caller zeroes the returned bytes.
  static func kek(slot: String) throws -> [UInt8] {
    var q = query(slot)
    q[kSecReturnData as String] = true
    q[kSecMatchLimit as String] = kSecMatchLimitOne
    var item: CFTypeRef?
    let status = SecItemCopyMatching(q as CFDictionary, &item)
    if status == errSecSuccess, let data = item as? Data {
      guard data.count == 32 else {
        throw Exception(name: "ERR_AEGIS_VAULT_KEK_LOST", description: "stored KEK is malformed")
      }
      return [UInt8](data)
    }
    guard status == errSecItemNotFound else {
      throw Exception(name: "ERR_AEGIS_VAULT", description: "Keychain read failed: \(status)")
    }
    var kek = [UInt8](repeating: 0, count: 32)
    guard SecRandomCopyBytes(kSecRandomDefault, kek.count, &kek) == errSecSuccess else {
      throw Exception(name: "ERR_AEGIS_VAULT", description: "no randomness for the KEK")
    }
    var add = query(slot)
    add[kSecValueData as String] = Data(kek)
    add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
    let added = SecItemAdd(add as CFDictionary, nil)
    guard added == errSecSuccess else {
      kek.withUnsafeMutableBytes { _ = memset_s($0.baseAddress, $0.count, 0, $0.count) }
      throw Exception(name: "ERR_AEGIS_VAULT", description: "Keychain write failed: \(added)")
    }
    return kek
  }

  /// Forget the profile's KEK: its blobs can never be unwrapped again (cryptographic erase).
  static func destroy(slot: String) throws {
    let status = SecItemDelete(query(slot) as CFDictionary)
    guard status == errSecSuccess || status == errSecItemNotFound else {
      throw Exception(name: "ERR_AEGIS_VAULT", description: "Keychain delete failed: \(status)")
    }
  }

  /// Slot ids are short ASCII names ('self', 'slot_1').
  static func checkSlot(_ slot: String) throws -> [UInt8] {
    guard slot.range(of: "^[A-Za-z0-9_.-]{1,64}$", options: .regularExpression) != nil else {
      throw Exception(name: "ERR_AEGIS_VAULT", description: "bad vault slot name")
    }
    return Array(slot.utf8)
  }
}
